File: server/typer/engine.go
package typer

import (
	"encoding/json"
	"strings"
	"unicode/utf8"

	"cloud-clicker/server/copykeys"
	"cloud-clicker/server/determinism"
	"cloud-clicker/server/minigame"
)

const (
	EngineRef          = "typer"
	RunSubstream       = "typer.run.v1"
	PromptSubstream    = "typer.prompts.v1"
	ScalingDestination = "typer.era_tier"

	PhaseReady    = "ready"
	PhaseTyping   = "typing"
	PhaseTerminal = "terminal"

	AssistTimed   = "timed"
	AssistUntimed = "untimed"

	OutcomeCompleted  = "completed"
	OutcomeEndedEarly = "ended_early"
	OutcomeTimedOut   = "timed_out"

	SubmissionCleared = "cleared"
	SubmissionMiss    = "miss"
)

var errorTaxonomy = []string{"illegal_phase", "invalid_assist_level", "invalid_text", "line_too_long"}

// resultFactKinds is the byte-sorted certified fact set (TT4.7).
var resultFactKinds = []string{"typer.assisted", "typer.clean_lines", "typer.elapsed_ms", "typer.lines_cleared", "typer.misses"}

// Submission fields are declared in byte order so the nested object is
// canonical (the platform re-canonicalizes every level).
type Submission struct {
	FirstMismatchIndex *int64 `json:"first_mismatch_index"`
	Outcome            string `json:"outcome"`
}

// Snapshot is typer.snapshot.v1: exactly eighteen keys (TT4.5).
type Snapshot struct {
	TyperContentHash    string      `json:"typer_content_hash"`
	TyperSchemaVersion  int         `json:"typer_schema_version"`
	Phase               string      `json:"phase"`
	EraTier             int64       `json:"era_tier"`
	AssistLevel         *string     `json:"assist_level"`
	PromptIndex         int64       `json:"prompt_index"`
	PromptsTotal        int64       `json:"prompts_total"`
	CurrentPromptID     *string     `json:"current_prompt_id"`
	CurrentPromptText   *string     `json:"current_prompt_text"`
	CurrentPromptMisses int64       `json:"current_prompt_misses"`
	LinesCleared        int64       `json:"lines_cleared"`
	CleanLines          int64       `json:"clean_lines"`
	Misses              int64       `json:"misses"`
	StartedServerMS     *int64      `json:"started_server_ms"`
	DeadlineServerMS    *int64      `json:"deadline_server_ms"`
	LastServerMS        *int64      `json:"last_server_ms"`
	LastSubmission      *Submission `json:"last_submission"`
	Revision            int64       `json:"revision"`
}

type Tenant struct{}

func NewTenant() Tenant { return Tenant{} }

func (Tenant) Descriptor() minigame.Descriptor {
	return minigame.Descriptor{EngineRef: EngineRef, EngineVersion: EngineVersion,
		CommandSchema: "typer.command.v1", SnapshotSchema: "typer.snapshot.v1", ResultSchema: "minigame.result.v1",
		Modes: []minigame.Mode{minigame.ModeSolo}, ErrorTaxonomy: append([]string(nil), errorTaxonomy...),
		Destinations: map[string]minigame.DestinationClass{ScalingDestination: minigame.DestinationBreadth}}
}

func (Tenant) ValidateCommand(data json.RawMessage) error {
	_, rejection := decodeCommand(data)
	return rejection
}

func (Tenant) ValidateSnapshot(data json.RawMessage) error {
	_, err := decodeSnapshot(data)
	return err
}

// ValidateResult is content-free: it checks the closed outcome and fact sets
// and the structural bounds. run_length and hardcaps are enforced by the
// engine, which holds the content.
func (Tenant) ValidateResult(result *minigame.Result) error {
	if result == nil {
		return nil
	}
	if result.Outcome != OutcomeCompleted && result.Outcome != OutcomeEndedEarly && result.Outcome != OutcomeTimedOut ||
		result.RatingDelta != nil || len(result.ScoreFacts) != len(resultFactKinds) {
		return minigame.ErrInvalidTenant
	}
	values := map[string]int64{}
	for index, fact := range result.ScoreFacts {
		if fact.Kind != resultFactKinds[index] || fact.Value < 0 {
			return minigame.ErrInvalidTenant
		}
		values[fact.Kind] = fact.Value
	}
	if values["typer.assisted"] > 1 || values["typer.clean_lines"] > values["typer.lines_cleared"] {
		return minigame.ErrInvalidTenant
	}
	return nil
}

func (Tenant) Create(input minigame.CreateInput) (json.RawMessage, error) {
	catalog, err := catalogForInput(input.Content, input.ContentHash, input.ContentSchemaVersion)
	if err != nil || input.Mode != minigame.ModeSolo {
		return nil, minigame.ErrInvalidTenant
	}
	eraTier, ok := validScaling(input.ScalingInputs, catalog)
	if !ok {
		return nil, minigame.ErrInvalidTenant
	}
	return encodeSnapshot(Snapshot{TyperContentHash: input.ContentHash, TyperSchemaVersion: input.ContentSchemaVersion,
		Phase: PhaseReady, EraTier: eraTier, PromptsTotal: catalog.Policy.RunLength, Revision: 1})
}

func (Tenant) Apply(input minigame.ApplyInput) (minigame.ApplyOutput, error) {
	catalog, err := catalogForInput(input.Content, input.ContentHash, input.ContentSchemaVersion)
	if err != nil || input.Mode != minigame.ModeSolo {
		return minigame.ApplyOutput{}, minigame.ErrInvalidTenant
	}
	eraTier, ok := validScaling(input.ScalingInputs, catalog)
	if !ok {
		return minigame.ApplyOutput{}, minigame.ErrInvalidTenant
	}
	snapshot, err := decodeSnapshot(input.Snapshot)
	if err != nil || snapshot.Revision != input.Revision || snapshot.TyperContentHash != input.ContentHash ||
		snapshot.TyperSchemaVersion != input.ContentSchemaVersion || snapshot.EraTier != eraTier ||
		validateSnapshotAgainstCatalog(snapshot, catalog, input.Seed) != nil {
		return minigame.ApplyOutput{}, minigame.ErrTenantDivergence
	}
	command, rejection := decodeCommand(input.Command)
	if rejection != nil {
		return minigame.ApplyOutput{}, rejection
	}
	result, rejection := transition(&snapshot, command, catalog, input.Seed, input.ServerTimeMs)
	if rejection != nil {
		return minigame.ApplyOutput{}, rejection
	}
	snapshot.Revision = input.Revision + 1
	encoded, err := encodeSnapshot(snapshot)
	if err != nil {
		return minigame.ApplyOutput{}, minigame.ErrTenantDivergence
	}
	return minigame.ApplyOutput{Snapshot: encoded, Result: result}, nil
}

func transition(snapshot *Snapshot, command command, catalog *Catalog, seed uint64, serverTimeMS int64) (*minigame.Result, error) {
	switch command.Kind {
	case "begin":
		if snapshot.Phase != PhaseReady {
			return nil, reject("illegal_phase", "begin requires ready phase")
		}
		t := advanceClock(snapshot, serverTimeMS)
		assist := command.AssistLevel
		snapshot.AssistLevel = &assist
		snapshot.StartedServerMS = int64Pointer(t)
		if assist == AssistTimed {
			snapshot.DeadlineServerMS = int64Pointer(t + catalog.Policy.TimedBudgetMS)
		}
		snapshot.Phase = PhaseTyping
		setCurrentPrompt(snapshot, PromptOrder(catalog, seed, snapshot.EraTier)[0])
		return nil, nil
	case "submit_line":
		if snapshot.Phase != PhaseTyping {
			return nil, reject("illegal_phase", "submit_line requires typing phase")
		}
		if int64(len(command.Text)) > catalog.Policy.MaxLineBytes {
			return nil, reject("line_too_long", "submitted line exceeds max_line_bytes")
		}
		t := advanceClock(snapshot, serverTimeMS)
		if snapshot.DeadlineServerMS != nil && t > *snapshot.DeadlineServerMS {
			return terminate(snapshot, catalog, OutcomeTimedOut), nil
		}
		normalized := Normalize(command.Text)
		target := *snapshot.CurrentPromptText
		if normalized == target {
			snapshot.LinesCleared++
			if snapshot.CurrentPromptMisses == 0 {
				snapshot.CleanLines++
			}
			snapshot.PromptIndex++
			snapshot.CurrentPromptMisses = 0
			snapshot.LastSubmission = &Submission{Outcome: SubmissionCleared}
			if snapshot.PromptIndex == catalog.Policy.RunLength {
				return terminate(snapshot, catalog, OutcomeCompleted), nil
			}
			setCurrentPrompt(snapshot, PromptOrder(catalog, seed, snapshot.EraTier)[snapshot.PromptIndex])
			return nil, nil
		}
		index := FirstMismatchIndex(normalized, target)
		snapshot.Misses = saturatingIncrement(snapshot.Misses, catalog.Policy.MissesHardcap)
		snapshot.CurrentPromptMisses = saturatingIncrement(snapshot.CurrentPromptMisses, catalog.Policy.MissesHardcap)
		snapshot.LastSubmission = &Submission{Outcome: SubmissionMiss, FirstMismatchIndex: &index}
		return nil, nil
	case "end_run":
		if snapshot.Phase != PhaseReady && snapshot.Phase != PhaseTyping {
			return nil, reject("illegal_phase", "end_run requires a non-terminal phase")
		}
		advanceClock(snapshot, serverTimeMS)
		return terminate(snapshot, catalog, OutcomeEndedEarly), nil
	default:
		return nil, reject("illegal_phase", "unknown command kind")
	}
}

// advanceClock applies t = max(sample, last_server_ms): a clock step
// backwards never rewinds the run.
func advanceClock(snapshot *Snapshot, sample int64) int64 {
	t := sample
	if snapshot.LastServerMS != nil && *snapshot.LastServerMS > t {
		t = *snapshot.LastServerMS
	}
	snapshot.LastServerMS = int64Pointer(t)
	return t
}

func setCurrentPrompt(snapshot *Snapshot, prompt Prompt) {
	id, text := prompt.PromptID, prompt.Text
	snapshot.CurrentPromptID, snapshot.CurrentPromptText = &id, &text
}

func terminate(snapshot *Snapshot, catalog *Catalog, outcome string) *minigame.Result {
	snapshot.Phase = PhaseTerminal
	snapshot.CurrentPromptID, snapshot.CurrentPromptText = nil, nil
	return &minigame.Result{Outcome: outcome, RatingDelta: nil, ScoreFacts: []minigame.ScoreFact{
		{Kind: "typer.assisted", Value: assistedFact(snapshot)},
		{Kind: "typer.clean_lines", Value: snapshot.CleanLines},
		{Kind: "typer.elapsed_ms", Value: elapsedFact(snapshot, catalog)},
		{Kind: "typer.lines_cleared", Value: snapshot.LinesCleared},
		{Kind: "typer.misses", Value: snapshot.Misses},
	}}
}

func assistedFact(snapshot *Snapshot) int64 {
	if snapshot.AssistLevel != nil && *snapshot.AssistLevel == AssistUntimed {
		return 1
	}
	return 0
}

func elapsedFact(snapshot *Snapshot, catalog *Catalog) int64 {
	if snapshot.StartedServerMS == nil || snapshot.LastServerMS == nil {
		return 0
	}
	end := *snapshot.LastServerMS
	if snapshot.DeadlineServerMS != nil && *snapshot.DeadlineServerMS < end {
		end = *snapshot.DeadlineServerMS
	}
	elapsed := end - *snapshot.StartedServerMS
	if elapsed > catalog.Policy.ElapsedHardcapMS {
		elapsed = catalog.Policy.ElapsedHardcapMS
	}
	return elapsed
}

func saturatingIncrement(value, hardcap int64) int64 {
	if value >= hardcap {
		return hardcap
	}
	return value + 1
}

// Normalize applies OD-7's closed smart-punctuation map and nothing else.
func Normalize(text string) string {
	return strings.NewReplacer("‘", "'", "’", "'", "“", "\"", "”", "\"", " ", " ", "—", "--").Replace(text)
}

// FirstMismatchIndex is the first differing byte offset, or the shorter
// length when one is a prefix of the other.
func FirstMismatchIndex(submitted, target string) int64 {
	limit := len(submitted)
	if len(target) < limit {
		limit = len(target)
	}
	for index := 0; index < limit; index++ {
		if submitted[index] != target[index] {
			return int64(index)
		}
	}
	return int64(limit)
}

// PromptOrder is TT4.2: a downward Fisher–Yates over the eligible pool with
// the typer.prompts.v1 substream of the typer.run.v1 run seed.
func PromptOrder(catalog *Catalog, seed uint64, eraTier int64) []Prompt {
	pool := catalog.Pool(eraTier)
	runSeed := determinism.Substream(seed, RunSubstream).Next()
	random := determinism.Substream(runSeed, PromptSubstream)
	for index := len(pool) - 1; index > 0; index-- {
		swap := int(random.Bound(uint64(index + 1)))
		pool[index], pool[swap] = pool[swap], pool[index]
	}
	return pool[:catalog.Policy.RunLength]
}

type command struct {
	Kind        string
	AssistLevel string
	Text        string
}

func decodeCommand(data []byte) (command, error) {
	if !uniqueJSONKeys(data) {
		return command{}, reject("illegal_phase", "command keys are not unique")
	}
	var header struct {
		Kind string `json:"kind"`
	}
	if json.Unmarshal(data, &header) != nil {
		return command{}, reject("illegal_phase", "command is not an object")
	}
	switch header.Kind {
	case "begin":
		var wire struct {
			Kind        string `json:"kind"`
			AssistLevel string `json:"assist_level"`
		}
		if !hasExactJSONKeys(data, "kind", "assist_level") || strictDecode(data, &wire) != nil ||
			wire.AssistLevel != AssistTimed && wire.AssistLevel != AssistUntimed {
			return command{}, reject("invalid_assist_level", "begin requires assist_level timed or untimed")
		}
		return command{Kind: header.Kind, AssistLevel: wire.AssistLevel}, nil
	case "submit_line":
		var wire struct {
			Kind string `json:"kind"`
			Text string `json:"text"`
		}
		if !hasExactJSONKeys(data, "kind", "text") || strictDecode(data, &wire) != nil {
			return command{}, reject("invalid_text", "submit_line schema mismatch")
		}
		if !validSubmittedText(wire.Text) {
			return command{}, reject("invalid_text", "submitted line contains invalid characters")
		}
		return command{Kind: header.Kind, Text: wire.Text}, nil
	case "end_run":
		if !hasExactJSONKeys(data, "kind") {
			return command{}, reject("illegal_phase", "end_run schema mismatch")
		}
		return command{Kind: header.Kind}, nil
	default:
		return command{}, reject("illegal_phase", "unknown command kind")
	}
}

// validSubmittedText is TT4.4 step 1: valid UTF-8 with no C0 control
// characters or U+007F.
func validSubmittedText(text string) bool {
	if !utf8.ValidString(text) {
		return false
	}
	for _, character := range text {
		if character < 0x20 || character == 0x7f || character == utf8.RuneError {
			return false
		}
	}
	return true
}

func decodeSnapshot(data []byte) (Snapshot, error) {
	if !uniqueJSONKeys(data) || !hasExactJSONKeys(data, "typer_content_hash", "typer_schema_version", "phase", "era_tier", "assist_level",
		"prompt_index", "prompts_total", "current_prompt_id", "current_prompt_text", "current_prompt_misses", "lines_cleared", "clean_lines",
		"misses", "started_server_ms", "deadline_server_ms", "last_server_ms", "last_submission", "revision") {
		return Snapshot{}, minigame.ErrInvalidTenant
	}
	var value Snapshot
	if strictDecode(data, &value) != nil || value.TyperSchemaVersion != SchemaVersion ||
		!strings.HasPrefix(value.TyperContentHash, "sha256:") || len(value.TyperContentHash) != 71 ||
		value.Phase != PhaseReady && value.Phase != PhaseTyping && value.Phase != PhaseTerminal ||
		value.EraTier < 0 || value.EraTier > EraTierMax || value.Revision < 1 || value.PromptsTotal < 1 ||
		value.PromptIndex < 0 || value.PromptIndex > value.PromptsTotal || value.CurrentPromptMisses < 0 ||
		value.LinesCleared != value.PromptIndex || value.CleanLines < 0 || value.CleanLines > value.LinesCleared || value.Misses < 0 ||
		value.AssistLevel != nil && *value.AssistLevel != AssistTimed && *value.AssistLevel != AssistUntimed ||
		(value.CurrentPromptID == nil) != (value.CurrentPromptText == nil) ||
		(value.Phase == PhaseTyping) != (value.CurrentPromptID != nil) ||
		(value.Phase == PhaseReady) != (value.AssistLevel == nil && value.StartedServerMS == nil && value.LastServerMS == nil && value.PromptIndex == 0) && value.Phase != PhaseTerminal ||
		(value.AssistLevel == nil) != (value.StartedServerMS == nil) ||
		(value.DeadlineServerMS != nil) != (value.AssistLevel != nil && *value.AssistLevel == AssistTimed) {
		return Snapshot{}, minigame.ErrInvalidTenant
	}
	if value.LastSubmission != nil && (!hasExactSubmissionShape(value.LastSubmission)) {
		return Snapshot{}, minigame.ErrInvalidTenant
	}
	return value, nil
}

func hasExactSubmissionShape(value *Submission) bool {
	switch value.Outcome {
	case SubmissionCleared:
		return value.FirstMismatchIndex == nil
	case SubmissionMiss:
		return value.FirstMismatchIndex != nil && *value.FirstMismatchIndex >= 0
	default:
		return false
	}
}

func validateSnapshotAgainstCatalog(value Snapshot, catalog *Catalog, seed uint64) error {
	if value.PromptsTotal != catalog.Policy.RunLength || value.Misses > catalog.Policy.MissesHardcap ||
		value.CurrentPromptMisses > catalog.Policy.MissesHardcap {
		return minigame.ErrInvalidTenant
	}
	if value.Phase == PhaseTyping {
		expected := PromptOrder(catalog, seed, value.EraTier)[value.PromptIndex]
		if *value.CurrentPromptID != expected.PromptID || *value.CurrentPromptText != expected.Text {
			return minigame.ErrInvalidTenant
		}
	}
	return nil
}

func encodeSnapshot(snapshot Snapshot) (json.RawMessage, error) {
	encoded, err := json.Marshal(snapshot)
	if err != nil {
		return nil, err
	}
	var canonical map[string]json.RawMessage
	if json.Unmarshal(encoded, &canonical) != nil {
		return nil, minigame.ErrTenantDivergence
	}
	return json.Marshal(canonical)
}

func catalogForInput(data []byte, hash string, schemaVersion int) (*Catalog, error) {
	if len(data) == 0 || hash != ContentHash(data) || schemaVersion != SchemaVersion {
		return nil, ErrInvalidCatalog
	}
	keys := map[string]struct{}{}
	for _, key := range copykeys.All() {
		keys[key] = struct{}{}
	}
	return LoadCatalog(data, Declarations{CopyKeys: keys})
}

func validScaling(values map[string]int64, catalog *Catalog) (int64, bool) {
	if len(values) != 1 {
		return 0, false
	}
	tier, ok := values[ScalingDestination]
	if !ok || tier < 0 || tier > EraTierMax || int64(len(catalog.Pool(tier))) < catalog.Policy.RunLength {
		return 0, false
	}
	return tier, true
}

func int64Pointer(value int64) *int64 { return &value }

func reject(code, detail string) *minigame.Rejection {
	return &minigame.Rejection{Code: code, Detail: detail}
}

File: client/src/typer/engine.ts
import { substream } from "../combat/rng";
import { COPY_KEYS } from "../copy";
import { parseTyperCatalog, typerContentHash, typerPool, TYPER_ERA_TIER_MAX, TYPER_SCHEMA_VERSION, type TyperCatalog, type TyperPrompt } from "./catalog";

// Terminal Typer pure engine (TT4), the TS mirror of server/typer/engine.go.
export const TYPER_ENGINE_VERSION = "1.0.0" as const;
export const TYPER_SCALING_DESTINATION = "typer.era_tier" as const;
const RUN_SUBSTREAM = "typer.run.v1";
const PROMPT_SUBSTREAM = "typer.prompts.v1";

export type TyperPhase = "ready" | "typing" | "terminal";
export type TyperAssist = "timed" | "untimed";
export type TyperCommand = Readonly<{ kind: "begin"; assist_level: TyperAssist }> | Readonly<{ kind: "submit_line"; text: string }> | Readonly<{ kind: "end_run" }>;
export interface TyperSubmission { readonly outcome: "cleared" | "miss"; readonly first_mismatch_index: number | null }
export interface TyperSnapshot {
  readonly assist_level: TyperAssist | null; readonly clean_lines: number; readonly current_prompt_id: string | null;
  readonly current_prompt_misses: number; readonly current_prompt_text: string | null; readonly deadline_server_ms: number | null;
  readonly era_tier: number; readonly last_server_ms: number | null; readonly last_submission: TyperSubmission | null;
  readonly lines_cleared: number; readonly misses: number; readonly phase: TyperPhase; readonly prompt_index: number;
  readonly prompts_total: number; readonly revision: number; readonly started_server_ms: number | null;
  readonly typer_content_hash: string; readonly typer_schema_version: number;
}
export interface TyperResult {
  readonly outcome: "completed" | "ended_early" | "timed_out"; readonly rating_delta: null;
  readonly score_facts: readonly { readonly kind: string; readonly value: number }[];
}
export type TyperRejectionCode = "illegal_phase" | "invalid_assist_level" | "invalid_text" | "line_too_long";
export class TyperRejection extends Error { constructor(readonly code: TyperRejectionCode, readonly detail: string) { super(`${code}: ${detail}`); } }

export interface TyperContentInput { readonly content: string; readonly content_hash: string; readonly content_schema_version: number }
export interface TyperCreateInput extends TyperContentInput { readonly seed: bigint; readonly mode: "solo"; readonly scaling_inputs: Readonly<Record<string, number>> }
export interface TyperApplyInput extends TyperCreateInput { readonly revision: number; readonly snapshot: string; readonly command: string; readonly server_time_ms: number }

type Mutable = { -readonly [K in keyof TyperSnapshot]: TyperSnapshot[K] };

export async function createTyper(input: TyperCreateInput): Promise<string> {
  const catalog = await resolveCatalog(input);
  if (input.mode !== "solo") throw new SyntaxError("Typer is solo only");
  const eraTier = validScaling(input.scaling_inputs, catalog);
  return encodeSnapshot({ assist_level: null, clean_lines: 0, current_prompt_id: null, current_prompt_misses: 0, current_prompt_text: null,
    deadline_server_ms: null, era_tier: eraTier, last_server_ms: null, last_submission: null, lines_cleared: 0, misses: 0, phase: "ready",
    prompt_index: 0, prompts_total: catalog.policy.run_length, revision: 1, started_server_ms: null,
    typer_content_hash: input.content_hash, typer_schema_version: input.content_schema_version });
}

export async function applyTyper(input: TyperApplyInput): Promise<{ readonly snapshot: string; readonly result: TyperResult | null }> {
  const catalog = await resolveCatalog(input);
  const eraTier = validScaling(input.scaling_inputs, catalog);
  const snapshot = decodeSnapshot(input.snapshot);
  if (snapshot.revision !== input.revision || snapshot.typer_content_hash !== input.content_hash || snapshot.typer_schema_version !== input.content_schema_version ||
    snapshot.era_tier !== eraTier) throw new SyntaxError("Typer snapshot diverges from its identity");
  validateAgainstCatalog(snapshot, catalog, input.seed);
  const command = decodeCommand(input.command);
  const result = transition(snapshot, command, catalog, input.seed, input.server_time_ms);
  snapshot.revision = input.revision + 1;
  return { snapshot: encodeSnapshot(snapshot), result };
}

function transition(snapshot: Mutable, command: TyperCommand, catalog: TyperCatalog, seed: bigint, serverTimeMS: number): TyperResult | null {
  if (command.kind === "begin") {
    if (snapshot.phase !== "ready") throw new TyperRejection("illegal_phase", "begin requires ready phase");
    const t = advanceClock(snapshot, serverTimeMS);
    snapshot.assist_level = command.assist_level;
    snapshot.started_server_ms = t;
    if (command.assist_level === "timed") snapshot.deadline_server_ms = t + catalog.policy.timed_budget_ms;
    snapshot.phase = "typing";
    setCurrent(snapshot, typerPromptOrder(catalog, seed, snapshot.era_tier)[0]!);
    return null;
  }
  if (command.kind === "submit_line") {
    if (snapshot.phase !== "typing") throw new TyperRejection("illegal_phase", "submit_line requires typing phase");
    if (new TextEncoder().encode(command.text).length > catalog.policy.max_line_bytes) throw new TyperRejection("line_too_long", "submitted line exceeds max_line_bytes");
    const t = advanceClock(snapshot, serverTimeMS);
    if (snapshot.deadline_server_ms !== null && t > snapshot.deadline_server_ms) return terminate(snapshot, catalog, "timed_out");
    const normalized = normalizeTyperLine(command.text), target = snapshot.current_prompt_text!;
    if (normalized === target) {
      snapshot.lines_cleared++;
      if (snapshot.current_prompt_misses === 0) snapshot.clean_lines++;
      snapshot.prompt_index++;
      snapshot.current_prompt_misses = 0;
      snapshot.last_submission = { outcome: "cleared", first_mismatch_index: null };
      if (snapshot.prompt_index === catalog.policy.run_length) return terminate(snapshot, catalog, "completed");
      setCurrent(snapshot, typerPromptOrder(catalog, seed, snapshot.era_tier)[snapshot.prompt_index]!);
      return null;
    }
    const index = firstMismatchIndex(normalized, target);
    snapshot.misses = Math.min(snapshot.misses + 1, catalog.policy.misses_hardcap);
    snapshot.current_prompt_misses = Math.min(snapshot.current_prompt_misses + 1, catalog.policy.misses_hardcap);
    snapshot.last_submission = { outcome: "miss", first_mismatch_index: index };
    return null;
  }
  if (snapshot.phase !== "ready" && snapshot.phase !== "typing") throw new TyperRejection("illegal_phase", "end_run requires a non-terminal phase");
  advanceClock(snapshot, serverTimeMS);
  return terminate(snapshot, catalog, "ended_early");
}

function advanceClock(snapshot: Mutable, sample: number): number {
  const t = snapshot.last_server_ms !== null && snapshot.last_server_ms > sample ? snapshot.last_server_ms : sample;
  snapshot.last_server_ms = t;
  return t;
}

function setCurrent(snapshot: Mutable, prompt: TyperPrompt): void { snapshot.current_prompt_id = prompt.prompt_id; snapshot.current_prompt_text = prompt.text; }

function terminate(snapshot: Mutable, catalog: TyperCatalog, outcome: TyperResult["outcome"]): TyperResult {
  snapshot.phase = "terminal"; snapshot.current_prompt_id = null; snapshot.current_prompt_text = null;
  let elapsed = 0;
  if (snapshot.started_server_ms !== null && snapshot.last_server_ms !== null) {
    const end = snapshot.deadline_server_ms !== null && snapshot.deadline_server_ms < snapshot.last_server_ms ? snapshot.deadline_server_ms : snapshot.last_server_ms;
    elapsed = Math.min(end - snapshot.started_server_ms, catalog.policy.elapsed_hardcap_ms);
  }
  return { outcome, rating_delta: null, score_facts: [
    { kind: "typer.assisted", value: snapshot.assist_level === "untimed" ? 1 : 0 },
    { kind: "typer.clean_lines", value: snapshot.clean_lines },
    { kind: "typer.elapsed_ms", value: elapsed },
    { kind: "typer.lines_cleared", value: snapshot.lines_cleared },
    { kind: "typer.misses", value: snapshot.misses },
  ] };
}

const NORMALIZATION: readonly [string, string][] = [["‘", "'"], ["’", "'"], ["“", "\""], ["”", "\""], [" ", " "], ["—", "--"]];

/** OD-7's closed smart-punctuation map and nothing else. */
export function normalizeTyperLine(text: string): string {
  let result = "";
  for (const character of text) result += NORMALIZATION.find(([from]) => from === character)?.[1] ?? character;
  return result;
}

/** First differing UTF-8 byte offset, or the shorter length on a prefix. */
export function firstMismatchIndex(submitted: string, target: string): number {
  const left = new TextEncoder().encode(submitted), right = new TextEncoder().encode(target);
  const limit = Math.min(left.length, right.length);
  for (let index = 0; index < limit; index++) if (left[index] !== right[index]) return index;
  return limit;
}

/** TT4.2: downward Fisher–Yates with the typer.prompts.v1 substream of the run seed. */
export function typerPromptOrder(catalog: TyperCatalog, seed: bigint, eraTier: number): TyperPrompt[] {
  const pool = typerPool(catalog, eraTier);
  const runSeed = substream(seed, RUN_SUBSTREAM).next();
  const random = substream(runSeed, PROMPT_SUBSTREAM);
  for (let index = pool.length - 1; index > 0; index--) {
    const swap = Number(random.bound(BigInt(index + 1)));
    [pool[index], pool[swap]] = [pool[swap]!, pool[index]!];
  }
  return pool.slice(0, catalog.policy.run_length);
}

function decodeCommand(source: string): TyperCommand {
  let value: unknown;
  try { value = JSON.parse(source); } catch { throw new TyperRejection("illegal_phase", "command is not JSON"); }
  if (value === null || typeof value !== "object" || Array.isArray(value)) throw new TyperRejection("illegal_phase", "command is not an object");
  const row = value as Record<string, unknown>, keys = Object.keys(row).sort().join("\0");
  if (row.kind === "begin") {
    if (keys !== "assist_level\0kind" || row.assist_level !== "timed" && row.assist_level !== "untimed") throw new TyperRejection("invalid_assist_level", "begin requires assist_level timed or untimed");
    return { kind: "begin", assist_level: row.assist_level };
  }
  if (row.kind === "submit_line") {
    if (keys !== "kind\0text" || typeof row.text !== "string") throw new TyperRejection("invalid_text", "submit_line schema mismatch");
    if (!validSubmittedText(row.text)) throw new TyperRejection("invalid_text", "submitted line contains invalid characters");
    return { kind: "submit_line", text: row.text };
  }
  if (row.kind === "end_run") {
    if (keys !== "kind") throw new TyperRejection("illegal_phase", "end_run schema mismatch");
    return { kind: "end_run" };
  }
  throw new TyperRejection("illegal_phase", "unknown command kind");
}

// TT4.4 step 1. U+FFFD is refused in both runtimes because Go's JSON decoder
// substitutes it for invalid UTF-8; lone surrogates are invalid UTF-16 here.
function validSubmittedText(text: string): boolean {
  for (let index = 0; index < text.length; index++) {
    const code = text.charCodeAt(index);
    if (code < 0x20 || code === 0x7f || code === 0xfffd) return false;
    if (code >= 0xd800 && code <= 0xdbff) { const next = text.charCodeAt(index + 1); if (!(next >= 0xdc00 && next <= 0xdfff)) return false; index++; continue; }
    if (code >= 0xdc00 && code <= 0xdfff) return false;
  }
  return true;
}

function decodeSnapshot(source: string): Mutable {
  const value = JSON.parse(source) as Mutable;
  const keys = ["assist_level", "clean_lines", "current_prompt_id", "current_prompt_misses", "current_prompt_text", "deadline_server_ms", "era_tier",
    "last_server_ms", "last_submission", "lines_cleared", "misses", "phase", "prompt_index", "prompts_total", "revision", "started_server_ms",
    "typer_content_hash", "typer_schema_version"];
  if (Object.keys(value).sort().join("\0") !== keys.join("\0") || value.typer_schema_version !== TYPER_SCHEMA_VERSION ||
    !/^sha256:[0-9a-f]{64}$/.test(value.typer_content_hash) || !["ready", "typing", "terminal"].includes(value.phase) ||
    value.prompt_index < 0 || value.prompt_index > value.prompts_total || value.lines_cleared !== value.prompt_index ||
    value.clean_lines > value.lines_cleared || (value.phase === "typing") !== (value.current_prompt_id !== null) ||
    (value.assist_level === null) !== (value.started_server_ms === null) ||
    (value.deadline_server_ms !== null) !== (value.assist_level === "timed")) throw new SyntaxError("invalid Typer snapshot");
  return value;
}

function validateAgainstCatalog(snapshot: Mutable, catalog: TyperCatalog, seed: bigint): void {
  if (snapshot.prompts_total !== catalog.policy.run_length) throw new SyntaxError("Typer snapshot/catalog divergence");
  if (snapshot.phase === "typing") {
    const expected = typerPromptOrder(catalog, seed, snapshot.era_tier)[snapshot.prompt_index];
    if (!expected || expected.prompt_id !== snapshot.current_prompt_id || expected.text !== snapshot.current_prompt_text) throw new SyntaxError("Typer snapshot prompt divergence");
  }
}

/** Go-canonical JSON: sorted keys and Go's <, >, & escaping. */
export function encodeSnapshot(snapshot: TyperSnapshot): string {
  const sorted = (value: unknown): unknown => value === null || typeof value !== "object" ? value :
    Object.fromEntries(Object.keys(value as object).sort().map((key) => [key, sorted((value as Record<string, unknown>)[key])]));
  return JSON.stringify(sorted(snapshot)).replace(new RegExp("[<>&\\u2028\\u2029]", "gu"), (character) => `\\u${character.charCodeAt(0).toString(16).padStart(4, "0")}`);
}

async function resolveCatalog(input: TyperContentInput): Promise<TyperCatalog> {
  if (input.content_schema_version !== TYPER_SCHEMA_VERSION || input.content_hash !== await typerContentHash(input.content)) throw new SyntaxError("Typer content identity mismatch");
  return parseTyperCatalog(JSON.parse(input.content), new Set(COPY_KEYS));
}

function validScaling(values: Readonly<Record<string, number>>, catalog: TyperCatalog): number {
  const keys = Object.keys(values), tier = values[TYPER_SCALING_DESTINATION];
  if (keys.length !== 1 || tier === undefined || !Number.isSafeInteger(tier) || tier < 0 || tier > TYPER_ERA_TIER_MAX || typerPool(catalog, tier).length < catalog.policy.run_length) {
    throw new SyntaxError("invalid Typer scaling inputs");
  }
  return tier;
}
